packages/block/dom.ts

```diff
@@ -0,0 +1,177 @@
+export const setAttribute$ = element$.setAttribute;
+export const setAttributeNS$ = element$.setAttributeNS;
+export const setTextContent$ = Object.getOwnPropertyDescriptor(
+  node$,
+  'textContent',
+)!.set!;
+export const innerHTML$ = Object.getOwnPropertyDescriptor(
+  element$,
+  'innerHTML',
+)!.set!;
+export const childNodes$ = Object.getOwnPropertyDescriptor(node$, 'childNodes')!
+  .get!;
+export const characterDataSet$ = Object.getOwnPropertyDescriptor(
+  characterData$,
+  'data',
+)!.set!;
+
+export const map$ = Map.prototype;
+export const mapSet$ = map$.set;
+export const mapHas$ = map$.has;
+export const mapGet$ = map$.get;
+
+export const set$ = Set.prototype;
+export const setAdd$ = set$.add;
+export const setHas$ = set$.has;
+
+const eventCache = new Map<string, EventListener>();
+export const createEventListener = (
+  el: HTMLElement,
+  name: string,
+  value?: EventListener,
+) => {
+  const event = name.toLowerCase().slice(2);
+  if (!document[EVENTS_REGISTRY]) document[EVENTS_REGISTRY] = new Set();
+  if (!setHas$.call(document[EVENTS_REGISTRY], event)) {
+    // createEventListener uses a synthetic event handler to capture events
+    // https://betterprogramming.pub/whats-the-difference-between-synthetic-react-events-and-javascript-events-ba7dbc742294
+    addEventListener$.call(
+      document,
+      event,
+      (nativeEventObject: Event) => {
+        let el = nativeEventObject.target as Node | null;
+        // Bubble up the DOM tree to find all event listeners
+        while (el) {
+          const pool = el[EVENT_LISTENERS_POOL];
+          if (pool) {
+            const listeners = pool[event] as Record<string, any> | undefined;
+            if (listeners) {
+              const handlers = Object.values(listeners);
+              let returnFalse = false;
+              for (let i = 0, j = handlers.length; i < j; ++i) {
+                if (handlers[i](nativeEventObject) === false) {
+                  returnFalse = true;
+                }
+              }
+              if (returnFalse) return;
+            }
+          }
+          el = el.parentNode;
+        }
+      },
+      {
+        capture: false,
+      },
+    );
+    document[EVENTS_REGISTRY].add(event);
+  }
+  const pointer = listenerPointer++;
+  if (mapHas$.call(eventCache, value)) return mapGet$.call(eventCache, value);
+  const patch = (newValue?: EventListener) => {
+    if (!el[EVENT_LISTENERS_POOL]) el[EVENT_LISTENERS_POOL] = {};
+    const pool = el[EVENT_LISTENERS_POOL];
+    if (!pool[event]) pool[event] = {};
+    pool[event][pointer] = (e: Event) => {
+      (newValue ?? value)?.(e);
+    };
+  };
+  patch(value);
+  mapSet$.call(eventCache, value, patch);
+  return patch;
+};
+
+export const insertText = (
+  el: HTMLElement,
+  value: any,
+  index: number,
+): Text => {
+  const node = document.createTextNode(value);
+  const childNodes = childNodes$.call(el);
+  if (hasChildNodes$.call(el) && index < childNodes.length) {
```

Comment:
Fast path that can avoid a `hasChildNodes()` call:
```suggestion
  if (index < childNodes.length && hasChildNodes$.call(el)) {
```

Reply:
Also:

```
  const childNodes = childNodes$.call(el);
  if (hasChildNodes$.call(el) && index < childNodes.length) {
```

`childNodes` is accessed even before `hasChildNodes` is called, thus it is possible to eliminate that extra `childNodes` access.